pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1133,15 +1133,19 @@ impl<'a> DatadirModification<'a> {
         let writer = self.tline.writer().await;
 
         // Flush relation and  SLRU data blocks, keep metadata.
-        let mut result: anyhow::Result<()> = Ok(());
-        self.pending_updates.retain(|&key, value| {
-            if result.is_ok() && (is_rel_block_key(key) || is_slru_block_key(key)) {
-                result = writer.put(key, self.lsn, value);
-                false
-            } else {
-                true
+        let mut retain = HashSet::new();
+        let result = async {
+            for (key, value) in &self.pending_updates {
+                if is_rel_block_key(*key) || is_slru_block_key(*key) {
+                    writer.put(*key, self.lsn, value).await?;
+                } else {
+                    retain.insert(*key);
```

Comment:
I don't think this is equivalent to the original one. If we have a sequence key1, key2, key3, whereas key2 fails on put, the original code will retain key3, while this one will remove everything after the first error.

Reply:
😬 good catch, I implemented that in a hurry. Curious to see whether tests will fail.

Reply:
3fbe502a91a80659e214735feaf97644aebe3770